plenum/server/future_primaries_batch_handler.py

```diff
@@ -1,30 +1,87 @@
+        # Return delta of next primary's record
+        p_view_no = get_payload_data(txn)[AUDIT_TXN_VIEW_NO]
+        p_primaries = get_payload_data(txn)[AUDIT_TXN_PRIMARIES]
+        if p_view_no != view_no:
+            if isinstance(p_primaries, int):
+                return p_primaries
+            return 1
+        else:
+            if isinstance(p_primaries, int):
+                return p_primaries
+            return p_primaries
+
+    def _find_in_uncommitted(self, audit: Ledger, view_no, last_ind=1):
+        if len(audit.uncommittedTxns) == 0 or last_ind > len(audit.uncommittedTxns):
+            return None
+        last_txn = audit.uncommittedTxns[-last_ind]
+        primaries = self._inspect_audit_txn(last_txn, view_no)
+        if isinstance(primaries, list):
+            return primaries
+        return self._find_in_uncommitted(audit, view_no, last_ind + primaries)
+
+    def _get_previous_primaries(self, audit, view_no, seq_no_end):
+        if seq_no_end == 0:
+            return None
+        previous_txn = audit.getBySeqNo(seq_no_end)
+        primaries = self._inspect_audit_txn(previous_txn, view_no)
+        if isinstance(primaries, list):
+            return primaries
+        return self._get_previous_primaries(audit, view_no, seq_no_end - primaries)
+
+    def get_primaries_from_audit(self, view_no):
+        audit_ledger = self.db_manager.get_ledger(AUDIT_LEDGER_ID)
+        p_uncommited = self._find_in_uncommitted(audit_ledger, view_no)
+        if p_uncommited:
+            return p_uncommited
+        return self._get_previous_primaries(audit_ledger, view_no, audit_ledger.size)
+
+    def get_primaries(self, view_no):
+        if view_no not in self.primaries:
+            ps = self.get_primaries_from_audit(view_no)
+            if not ps:
+                return None
+            self.set_primaries(view_no, ps)
+        return self.primaries.get(view_no)
+
+    def check_primaries(self, view_no, primaries_to_check):
+        ps = self.get_primaries(view_no)
+        if ps:
+            return ps == primaries_to_check
+        return False
+
     def post_batch_applied(self, three_pc_batch: ThreePcBatch, prev_handler_result=None):
-        node_reg = list(self.node.nodeReg.keys())
-        number_of_inst = getMaxFailures(len(node_reg)) + 1
         view_no = self.node.viewNo if three_pc_batch.original_view_no is None else three_pc_batch.original_view_no
-        validators = TxnPoolManager.calc_node_names_ordered_by_rank(node_reg, copy.deepcopy(self.node.nodeIds))
-        three_pc_batch.primaries = self.node.primaries_selector.select_primaries(view_no=view_no,
-                                                                                 instance_count=number_of_inst,
-                                                                                 validators=validators)
+        primaries = self.get_primaries(view_no)
+        if primaries is None:
+            primaries = three_pc_batch.primaries
```

Comment:
Should it be `pre_prepare.primaries`?

Reply:
it would be pre_prepare.primaries because it will be set while three_pc_batch created.